build/linter/constructor/analyzer.go

```diff
@@ -0,0 +1,176 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package constructor
+
+import (
+	"go/ast"
+	"go/types"
+	"slices"
+	"strings"
+
+	"github.com/fatih/structtag"
+	"golang.org/x/tools/go/analysis"
+	"golang.org/x/tools/go/ast/inspector"
+)
+
+// Analyzer is the analyzer struct of constructor.
+// constructor only allows constructing a struct manually in some specific functions, which is specified with tags for
+// `constructor.Constructor` field.
+//
+// It can detect the following pattern and give error (if not in constructor functions):
+//
+// 1. Create struct directly, like `SomeStruct{}` or `&SomeStruct{}`.
+// 2. Create struct with `new`: `new(SomeStruct)`
+// 3. Struct literal in slice or other struct: `[]SomeStruct{{}}`
+// 4. Define variables through `var`: `var a SomeStruct`.
+// 5. Implicit zero value in other struct literal: `type OtherStruct struct{SomeStruct}; other := OtherStruct{}`
+//
+// TODO: verify whether this linter can work well with generics
+var Analyzer = &analysis.Analyzer{
+	Name:     "constructor",
+	Doc:      "Check developers don't create structs manually without using constructors",
+	Requires: []*analysis.Analyzer{},
+	Run:      run,
+}
+
+func getConstructorList(t types.Type) []string {
```

Comment:
I'm not sure whether (and how) I need to add a cache to this function. I don't know whether this linter will make the compilation significantly slower.

But I think it's not worse than the `noCopy` linter (see https://github.com/golang/tools/blob/12b5dadef96835a9e447f2079a6cb949220c3856/go/analysis/passes/copylock/copylock.go#L330)